JDK-8256287: [windows] add loop fuse to map_or_reserve_memory_aligned

tstuefe opened the pull request:
Hi,

may I please have reviews for this little fix:

On windows, `map_or_reserve_memory_aligned()` attempts to allocate aligned memory by:
1) reserving larger area
2) releasing it
3) attempting to re-reserve into the vacated address space a smaller area at the aligned start address

Since this may fail (between (2) and (3) someone may have grabbed part of that address space concurrently), we do this in a loop. However, when failing to release it we will loop-reserve endlessly.

This is one of the reasons for JDK-8255954 whose root cause will be fixed with JDK-8255978. But we should guard against endless loops independently from that.

This patch:
- limits the number we try this to 20. If in 20 cases someone else grabs address space concurrently, something is off...
- now correctly handles the return code of the release operation (1): in debug we assert, since if release_memory() fails something is wrong and we should take a look. In the release case, we now return an error.


Thanks!

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- |
| Build | ✔️ (5/5 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8256287](https://bugs.openjdk.java.net/browse/JDK-8256287): [windows] add loop fuse to map_or_reserve_memory_aligned


### Reviewers
 * [Ludovic Henry](https://openjdk.java.net/census#luhenry) (@luhenry - Author) ⚠️ Review applies to e716a994372b4e9d2c460acbf117677851144984
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)


### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1191/head:pull/1191`
`$ git checkout pull/1191`

shipilev commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3160,17 +3161,22 @@ static char* map_or_reserve_memory_aligned(size_t size, size_t alignment, int fi
-      os::unmap_memory(extra_base, extra_size);
-    } else {
-      os::release_memory(extra_base, extra_size);
+    bool rc = (file_desc != -1) ? os::unmap_memory(extra_base, extra_size) :
+                                  os::release_memory(extra_base, extra_size);
```
This style is inconsistent with style at L3155 and L3172. I think _those_ lines need to be brought in style.

shipilev commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3160,17 +3161,22 @@ static char* map_or_reserve_memory_aligned(size_t size, size_t alignment, int fi
+                                  os::release_memory(extra_base, extra_size);
+    assert(rc, "release failed");
+    if (!rc) {
+      aligned_base = NULL;
+      break;
```
Seems like we can just `return NULL`  on this path, like we do with `extra_base == NULL` check in the same loop? So we can have:

```
if (!rc) {
  assert(false, "Unmap/release should always succeed here");
  return NULL;
}
```

shipilev commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3149,6 +3149,7 @@ static char* map_or_reserve_memory_aligned(size_t size, size_t alignment, int fi
   assert(extra_size >= size, "overflow, size is too large to allow alignment");
 
   char* aligned_base = NULL;
+  int attempts = 20;
```
```
static const int max_attempts = 20;
int attempts = max_attempts;
```

...and then you can use `max_attempts` in the assert message?

shipilev commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3160,17 +3161,22 @@ static char* map_or_reserve_memory_aligned(size_t size, size_t alignment, int fi
     aligned_base = file_desc != -1 ?
       os::attempt_map_memory_to_file_at(aligned_base, size, file_desc) :
       os::attempt_reserve_memory_at(aligned_base, size);
+    attempts --;
```
No space before unary op? E.g. `attempts--`? It might be even cleaner to move it to loop condition like `attempts-- > 0`?

shipilev reviewed (commented):
This makes sense to me. Just a few drive-by stylistic comments.

tstuefe commented:
> This makes sense to me. Just a few drive-by stylistic comments.

Thanks Aleksey!

I updated the patch according your suggestions. Yes, that looks better. I also corrected the final assert for the case that the final remap attempt had been successful.

.. Thomas

luhenry commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -3149,28 +3149,34 @@ static char* map_or_reserve_memory_aligned(size_t size, size_t alignment, int fi
   assert(extra_size >= size, "overflow, size is too large to allow alignment");
 
   char* aligned_base = NULL;
+  static const int max_attempts = 20;
```
Stylistic review as well: I'd put that counter in a loop like `for (int attempts = 0; aligned_base == NULL && attempts < 20 /* max attemps */; attemps++) { ... }`. That would clearly indicate that it is a counted loop.

Otherwise, now that you can return `NULL` from this function, have you checked that all uses of this function deal with a NULL-returned value correctly? From a cursory glance, it looks good to me.

luhenry reviewed (approved):
I'm not a reviewer but it looks good to me.

tstuefe commented:
Hi Ludovic,

thanks a lot for looking at this! Reformed the loop as suggested.

> I'm not a reviewer but it looks good to me.

Your feedback is still useful. In the case of hotspot, the project rules want two reviews, only one of which has to be a Reviewer. 

I see that in the JDK project you are "just" an Author, but you are overdue for Committer anyway I think since you have ~11 patches authored or co-authored.

Cheers, Thomas

iklam reviewed (approved):
LGTM

yminqi reviewed (approved):
LGTM

tstuefe commented:
Thanks Calvin and Ioi!

tstuefe commented:
/integrate
